File: go-cli/internal/compose/compose.go
// Package compose ports cli/src/compose.ts: layered file copy, dependency
// merge, env var append, and anchor wiring — the actual project-composition
// algorithm shared by `new` and `add`.
package compose

import (
	"fmt"
	"path/filepath"
	"time"

	"hackpack/internal/fsutil"
	"hackpack/internal/registry"
	"hackpack/internal/types"
)

type Options struct {
	RegistryPath  string
	RegistryLabel string
	TargetDir     string
	ProjectName   string
	Base          string
	Features      []string
	Pages         []string
}

type Result struct {
	PostInstall []string
}

// mergeManifestDeps ports the same-named helper in compose.ts: JS bases merge
// into package.json, Python bases merge into pyproject.toml. A manifest just
// declares whichever field(s) apply to its language.
func mergeManifestDeps(targetDir string, manifest types.TemplateManifest) error {
	if manifest.Dependencies != nil || manifest.DevDependencies != nil || manifest.Scripts != nil {
		patch := map[string]interface{}{}
		if manifest.Dependencies != nil {
			patch["dependencies"] = manifest.Dependencies
		}
		if manifest.DevDependencies != nil {
			patch["devDependencies"] = manifest.DevDependencies
		}
		if manifest.Scripts != nil {
			patch["scripts"] = manifest.Scripts
		}
		if err := fsutil.MergeJsonFile(filepath.Join(targetDir, "package.json"), patch); err != nil {
			return err
		}
	}
	if len(manifest.PyDependencies) > 0 {
		if err := fsutil.MergeTomlDependencyArray(filepath.Join(targetDir, "pyproject.toml"), "dependencies", manifest.PyDependencies); err != nil {
			return err
		}
	}
	return nil
}

// copyLayeredFiles ports the same-named helper: shared files/ first, then an
// optional files-<base>/ overlay for framework-specific integration code.
func copyLayeredFiles(templateRoot, base, targetDir string, vars map[string]interface{}) error {
	shared := filepath.Join(templateRoot, "files")
	if fsutil.PathExists(shared) {
		if err := fsutil.CopyTemplateDir(shared, targetDir, vars); err != nil {
			return err
		}
	}
	overlay := filepath.Join(templateRoot, "files-"+base)
	if fsutil.PathExists(overlay) {
		if err := fsutil.CopyTemplateDir(overlay, targetDir, vars); err != nil {
			return err
		}
	}
	return nil
}

type ApplyFeatureResult struct {
	PostInstall []string
	Category    string
}

// ApplyFeature layers one feature's files/deps/env/wiring onto an existing
// project directory. Shared by compose() (new project) and the `add` command.
func ApplyFeature(registryPath, targetDir, base, featureName string, vars map[string]interface{}) (ApplyFeatureResult, error) {
	manifest, err := registry.LoadManifest(registryPath, "feature", featureName)
	if err != nil {
		return ApplyFeatureResult{}, err
	}
	if len(manifest.CompatibleWith) > 0 && !contains(manifest.CompatibleWith, base) {
		return ApplyFeatureResult{}, fmt.Errorf("feature %q is not compatible with base %q", featureName, base)
	}

	if err := copyLayeredFiles(registry.TemplateDir(registryPath, "feature", featureName), base, targetDir, vars); err != nil {
		return ApplyFeatureResult{}, err
	}
	if err := mergeManifestDeps(targetDir, manifest); err != nil {
		return ApplyFeatureResult{}, err
	}
	if len(manifest.EnvVars) > 0 {
		evs := make([]fsutil.EnvVarLike, len(manifest.EnvVars))
		for i, e := range manifest.EnvVars {
			evs[i] = fsutil.EnvVarLike{Key: e.Key, Value: e.Value, Comment: e.Comment}
		}
		if err := fsutil.AppendEnvVars(filepath.Join(targetDir, ".env.example"), evs); err != nil {
			return ApplyFeatureResult{}, err
		}
	}
	for _, w := range manifest.Wiring {
		if err := fsutil.InsertAtAnchor(filepath.Join(targetDir, w.File), w.Anchor, w.Insert); err != nil {
			return ApplyFeatureResult{}, err
		}
	}

	return ApplyFeatureResult{PostInstall: manifest.PostInstall, Category: manifest.Category}, nil
}

func contains(list []string, v string) bool {
	for _, s := range list {
		if s == v {
			return true
		}
	}
	return false
}

func Compose(opts Options) (Result, error) {
	vars := map[string]interface{}{"projectName": opts.ProjectName}
	var postInstall []string

	baseManifest, err := registry.LoadManifest(opts.RegistryPath, "base", opts.Base)
	if err != nil {
		return Result{}, err
	}
	if err := fsutil.CopyTemplateDir(registry.TemplateDir(opts.RegistryPath, "base", opts.Base), opts.TargetDir, vars); err != nil {
		return Result{}, err
	}
	postInstall = append(postInstall, baseManifest.PostInstall...)

	// installedCategories lets pages find "whichever auth feature was installed",
	// per the plan's "compose with selected auth feature, not a fixed default".
	installedCategories := map[string]string{}

	for _, feature := range opts.Features {
		result, err := ApplyFeature(opts.RegistryPath, opts.TargetDir, opts.Base, feature, vars)
		if err != nil {
			return Result{}, err
		}
		if result.Category != "" {
			installedCategories[result.Category] = feature
		}
		postInstall = append(postInstall, result.PostInstall...)
	}

	installedPages := []types.InstalledPage{}
	for _, page := range opts.Pages {
		manifest, err := registry.LoadManifest(opts.RegistryPath, "page", page)
		if err != nil {
			return Result{}, err
		}
		if len(manifest.CompatibleWith) > 0 && !contains(manifest.CompatibleWith, opts.Base) {
			return Result{}, fmt.Errorf("page %q is not compatible with base %q", page, opts.Base)
		}

		var variant *string
		if manifest.RequiresCategory != "" {
			v := installedCategories[manifest.RequiresCategory]
			if v == "" {
				v = "none"
			}
			variant = &v
		}

		pageRoot := registry.TemplateDir(opts.RegistryPath, "page", page)
		sourceRoot := pageRoot
		if manifest.HasVariants {
			variantName := "none"
			if variant != nil {
				variantName = *variant
			}
			sourceRoot = filepath.Join(pageRoot, "variants", variantName)
		}

		pageVars := map[string]interface{}{"projectName": opts.ProjectName, "entityName": page}
		if err := copyLayeredFiles(sourceRoot, opts.Base, opts.TargetDir, pageVars); err != nil {
			return Result{}, err
		}
		if err := mergeManifestDeps(opts.TargetDir, manifest); err != nil {
			return Result{}, err
		}
		for _, w := range manifest.Wiring {
			if err := fsutil.InsertAtAnchor(filepath.Join(opts.TargetDir, w.File), w.Anchor, w.Insert); err != nil {
				return Result{}, err
			}
		}
		installedPages = append(installedPages, types.InstalledPage{Name: page, Variant: variant})
	}

	now := time.Now().UTC().Format("2006-01-02T15:04:05.000Z")

	features := opts.Features
	if features == nil {
		features = []string{}
	}
	hackpackJSON := types.HackpackManifest{
		Base:      opts.Base,
		Features:  features,
		Pages:     installedPages,
		CreatedAt: now,
	}
	if err := fsutil.WriteJSONFile(filepath.Join(opts.TargetDir, "hackpack.json"), hackpackJSON); err != nil {
		return Result{}, err
	}

	lockFeatures := make([]types.LockRef, len(opts.Features))
	for i, f := range opts.Features {
		lockFeatures[i] = types.LockRef{Name: f, Source: opts.RegistryLabel}
	}
	lockPages := make([]types.LockPage, len(installedPages))
	for i, p := range installedPages {
		lockPages[i] = types.LockPage{Name: p.Name, Variant: p.Variant, Source: opts.RegistryLabel}
	}
	if lockFeatures == nil {
		lockFeatures = []types.LockRef{}
	}
	if lockPages == nil {
		lockPages = []types.LockPage{}
	}
	lock := types.HackpackLock{
		Registry:   opts.RegistryLabel,
		ResolvedAt: now,
		Base:       types.LockBase{Name: opts.Base, Source: opts.RegistryLabel},
		Features:   lockFeatures,
		Pages:      lockPages,
	}
	if err := fsutil.WriteJSONFile(filepath.Join(opts.TargetDir, "hackpack.lock"), lock); err != nil {
		return Result{}, err
	}

	return Result{PostInstall: dedupe(postInstall)}, nil
}

func dedupe(items []string) []string {
	seen := map[string]bool{}
	out := []string{}
	for _, i := range items {
		if !seen[i] {
			seen[i] = true
			out = append(out, i)
		}
	}
	return out
}

File: cli/src/compose.ts
import path from "node:path";
import fs from "node:fs/promises";
import { loadManifest, templateDir } from "./registry.ts";
import {
  copyTemplateDir,
  insertAtAnchor,
  mergeJsonFile,
  mergeTomlDependencyArray,
  appendEnvVars,
  pathExists,
} from "./fsutil.ts";
import type { HackpackManifest, HackpackLock, InstalledPage, TemplateManifest } from "./types.ts";

export interface ComposeOptions {
  registryPath: string;
  registryLabel: string;
  targetDir: string;
  projectName: string;
  base: string;
  features: string[];
  pages: string[];
  author?: string;
  description?: string;
  license?: "mit" | "apache-2.0" | "none";
  packageManager?: "npm" | "pnpm" | "yarn" | "bun";
}

export interface ComposeResult {
  postInstall: string[];
}

/** JS bases merge into package.json; Python bases merge into requirements.txt.
 * A manifest just declares whichever field(s) apply to its language. */
async function mergeManifestDeps(targetDir: string, manifest: TemplateManifest): Promise<void> {
  if (manifest.dependencies || manifest.devDependencies || manifest.scripts) {
    await mergeJsonFile(path.join(targetDir, "package.json"), {
      dependencies: manifest.dependencies,
      devDependencies: manifest.devDependencies,
      scripts: manifest.scripts,
    });
  }
  if (manifest.pyDependencies?.length) {
    await mergeTomlDependencyArray(path.join(targetDir, "pyproject.toml"), "dependencies", manifest.pyDependencies);
  }
}

/** Copies a template's file tree onto the project: shared files/ first, then an
 * optional files-<base>/ overlay for framework-specific integration code (e.g. how
 * an auth feature mounts its routes differs between Next.js, SvelteKit, Hono...). */
async function copyLayeredFiles(
  templateRoot: string,
  base: string,
  targetDir: string,
  vars: Record<string, string>,
): Promise<void> {
  const shared = path.join(templateRoot, "files");
  if (await pathExists(shared)) await copyTemplateDir(shared, targetDir, vars);
  const overlay = path.join(templateRoot, `files-${base}`);
  if (await pathExists(overlay)) await copyTemplateDir(overlay, targetDir, vars);
}

/** Layers one feature's files/deps/env/wiring onto an existing project directory.
 * Shared by `compose()` (new project) and the `add` command (post-init). */
export async function applyFeature(
  registryPath: string,
  targetDir: string,
  base: string,
  featureName: string,
  vars: Record<string, string>,
): Promise<{ postInstall: string[]; category?: string }> {
  const manifest = await loadManifest(registryPath, "feature", featureName);
  if (manifest.compatibleWith?.length && !manifest.compatibleWith.includes(base)) {
    throw new Error(`Feature "${featureName}" is not compatible with base "${base}"`);
  }

  await copyLayeredFiles(templateDir(registryPath, "feature", featureName), base, targetDir, vars);
  await mergeManifestDeps(targetDir, manifest);
  if (manifest.envVars?.length) {
    await appendEnvVars(path.join(targetDir, ".env.example"), manifest.envVars);
  }
  for (const w of manifest.wiring ?? []) {
    await insertAtAnchor(path.join(targetDir, w.file), w.anchor, w.insert);
  }

  return { postInstall: manifest.postInstall ?? [], category: manifest.category };
}

export async function compose(opts: ComposeOptions): Promise<ComposeResult> {
  const { registryPath, targetDir, projectName, base, features, pages } = opts;
  const author = opts.author ?? "";
  const description = opts.description ?? "";
  const license = opts.license ?? "mit";
  const packageManager = opts.packageManager ?? "npm";
  const vars = {
    projectName,
    author,
    description,
    year: String(new Date().getFullYear()),
    pm: packageManager,
  };
  const postInstall: string[] = [];

  // Universal hygiene layer (README, .editorconfig, .env.example baseline) —
  // copied before the base so a base can still override a file by shipping
  // its own same-named file (none currently do).
  const commonDir = path.join(registryPath, "_common");
  if (await pathExists(commonDir)) {
    await copyTemplateDir(commonDir, targetDir, vars);
  }
  if (license !== "none") {
    const licenseFile = license === "apache-2.0" ? "LICENSE-APACHE-2.0.hbs" : "LICENSE-MIT.hbs";
    const licenseSrc = path.join(commonDir, licenseFile);
    if (await pathExists(licenseSrc)) {
      const raw = await fs.readFile(licenseSrc, "utf-8");
      const Handlebars = (await import("handlebars")).default;
      await fs.writeFile(path.join(targetDir, "LICENSE"), Handlebars.compile(raw)(vars));
    }
  }

  const baseManifest = await loadManifest(registryPath, "base", base);
  await copyTemplateDir(templateDir(registryPath, "base", base), targetDir, vars);
  postInstall.push(...(baseManifest.postInstall ?? []));

  // categories map lets pages find "whichever auth feature was installed",
  // per the plan's "compose with selected auth feature, not a fixed default".
  const installedCategories: Record<string, string> = {};

  for (const feature of features) {
    const result = await applyFeature(registryPath, targetDir, base, feature, vars);
    if (result.category) installedCategories[result.category] = feature;
    postInstall.push(...result.postInstall);
  }

  const installedPages: InstalledPage[] = [];
  for (const page of pages) {
    const manifest = await loadManifest(registryPath, "page", page);
    if (manifest.compatibleWith?.length && !manifest.compatibleWith.includes(base)) {
      throw new Error(`Page "${page}" is not compatible with base "${base}"`);
    }
    const variant = manifest.requiresCategory
      ? (installedCategories[manifest.requiresCategory] ?? "none")
      : null;
    const pageRoot = templateDir(registryPath, "page", page);
    const sourceRoot = manifest.hasVariants ? path.join(pageRoot, "variants", variant ?? "none") : pageRoot;
    await copyLayeredFiles(sourceRoot, base, targetDir, { ...vars, entityName: page });

    await mergeManifestDeps(targetDir, manifest);
    for (const w of manifest.wiring ?? []) {
      await insertAtAnchor(path.join(targetDir, w.file), w.anchor, w.insert);
    }
    installedPages.push({ name: page, variant });
  }

  const hackpackJson: HackpackManifest = {
    base,
    features,
    pages: installedPages,
    createdAt: new Date().toISOString(),
    author: author || undefined,
    description: description || undefined,
    license,
    packageManager,
  };
  await fs.writeFile(path.join(targetDir, "hackpack.json"), JSON.stringify(hackpackJson, null, 2) + "\n");

  const lock: HackpackLock = {
    registry: opts.registryLabel,
    resolvedAt: new Date().toISOString(),
    base: { name: base, source: opts.registryLabel },
    features: features.map((name) => ({ name, source: opts.registryLabel })),
    pages: installedPages.map((p) => ({ ...p, source: opts.registryLabel })),
  };
  await fs.writeFile(path.join(targetDir, "hackpack.lock"), JSON.stringify(lock, null, 2) + "\n");

  return { postInstall: [...new Set(postInstall)] };
}
